### PR TITLE
Batchnorm in ATen

### PR DESCRIPTION
**Pull request by ezyang**
This commit moves BatchNorm derivatives into ATen, eliminating torch/csrc/autograd/functions/batch_normalization.cpp

Some refactoring along the way:
- Functions got renamed to remove `_forward` from their names
- CuDNN batchnorm forward was modified to return save_mean/save_std instead of take it as parameters.
- THNN batch normalization takes care of resizing `save_mean` and `save_std` on forward.

There are some shenanigans re batchnorm backwards in eval mode. I'm tracking that in https://github.com/pytorch/pytorch/issues/4284

### CODE REVIEW

File: aten/src/ATen/native/BatchNorm.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+#if AT_CUDNN_ENABLED()
+#include "THC/THC.h"
+#include "ATen/cudnn/cudnn-wrapper.h"
+#endif
+
+namespace at { namespace native {
+
+namespace {
+  void check_dims_match_num_input_features(const std::string& arg_name, int64_t expected, int64_t actual){
```
Comment (colesbury) [inline aten/src/ATen/native/BatchNorm.cpp]: `const char* arg_name`

(I realize this was in the previous code)

**Comment by ezyang** [comment]
I had to reduce the number of MAX_JOBS on Travis because it was OOMing.

File: aten/src/THCUNN/generic/BatchNormalization.cu [hunk 2]
```diff
@@ -38,6 +38,8 @@ void THNN_(BatchNormalization_updateOutput)(
   bool train, double momentum, double eps) {
 
   THCTensor_(resizeAs)(state, output_, input_);
+  THCTensor_(resizeAs)(state, saveMean_, runningMean_);
```
Comment (colesbury) [inline aten/src/THCUNN/generic/BatchNormalization.cu]: Should this be guarded by an `if (train)` like in BatchNormalization.c

File: aten/src/ATen/native/cudnn/BatchNorm.cpp [hunk 3]
```diff
@@ -161,6 +175,7 @@ std::tuple<Tensor, Tensor, Tensor> cudnn_batch_norm_backward(
   checkSameSize(c, input, grad_output);
   auto num_features = input->size(1);
   for (auto t : {weight, save_mean, save_var}) {
+    if (!t->defined()) continue;
```
Comment (colesbury) [inline aten/src/ATen/native/cudnn/BatchNorm.cpp]: It looks like these are always defined.

File: aten/src/ATen/native/cudnn/BatchNorm.cpp [hunk 4]
```diff
@@ -146,7 +155,12 @@ std::tuple<Tensor, Tensor, Tensor> cudnn_batch_norm_backward(
   CheckedFrom c = "cudnn_batch_norm_backward";
   setCuDNNStreamToCurrent();
 
-  checkAllDefined(c, {input, grad_output, weight, save_mean, save_var});
+  checkAllDefined(c, {input, grad_output, weight});
+  if (training) {
+    checkAllDefined(c, {save_mean, save_var});
+  } else {
+    throw std::runtime_error("cudnn_batch_norm_backward: !training backwards doesn't seem to work");
```
Comment (colesbury) [inline aten/src/ATen/native/cudnn/BatchNorm.cpp]: The error message is a bit awkward. Since cuDNN doesn't support `training=false`, why take it as a parameter?

File: tools/autograd/templates/Functions.cpp [hunk 5]
```diff
@@ -712,12 +712,17 @@ std::tuple<Tensor, Tensor, Tensor> batchnorm_double_backward(
...
-    const Tensor & save_std, // not Variable
-    const Tensor & running_mean, // not Variable
-    const Tensor & running_var, // not Variable
+    const Tensor & save_mean_v,
+    const Tensor & save_std_v,
+    const Tensor & running_mean_v,
+    const Tensor & running_var_v,
     bool training) {
 
+  const Tensor& save_mean = static_cast<const Variable&>(save_mean_v).data();
```
Comment (colesbury) [inline tools/autograd/templates/Functions.cpp]: Are these guaranteed to be defined? Why are we unpacking them at all?

File: aten/src/ATen/native/cudnn/BatchNorm.cpp (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/native/cudnn/BatchNorm.cpp]: I inherited it from the old code, but I see now it makes sense to expunge it entirely.

File: tools/autograd/templates/Functions.cpp (hunk 5 above)
Comment (ezyang) [reply tools/autograd/templates/Functions.cpp]: The back story is I couldn't figure out if it was a supported operation to directly pass Tensors (not Variables) to an ATen derivative function, so I compromised with turning them all into Variables and unpacking them. `save_mean` and `save_std` used to be possibly undefined but that broke things, so they're now always defined (I need to fix some stuff elsewhere in this patch, as you have noticed).

**Comment by naba89** [comment]
Hi,

> This commit moves BatchNorm derivatives into ATen, eliminating torch/csrc/autograd/functions/batch_normalization.cpp

Eliminating `batch_normalization.cpp`, does it also mean that code related to batch_norm could be eliminated from `torch/csrc/autograd/functions/init.cpp` ? 

I am trying to understand the how to port some custom functions from csrc/autograd/functions to ATen.

Regards
Nabarun

**Comment by naba89** [comment]
Hi,

It's probably not directly related, but in the BatchNormalization_updateOutput function, the arguments have an underscore in them `input_`.  However, in nn_parse.py, the `fwd_function` shows as such:

> BatchNormalization_updateOutput(THNNState* state, THTensor* input, THTensor* output, THTensor* weight, THTensor* bias, THTensor* running_mean, THTensor* running_var, THTensor* save_mean, THTensor* save_std, bool train, double momentum, double eps)

The underscore disappears. I am porting some of my custom THCUNN functions to ATen, but I am getting error as `can't find binding for 'input_'` and the `fwd_function` also still retains the underscores. Any idea what I might be missing, or what are the steps to link the THCUNN implementations through ATen?

Regards
Nabarun
